server/src/main/java/org/opensearch/action/get/TransportGetAction.java

```diff
@@ -88,16 +91,37 @@ protected boolean resolveIndex(GetRequest request) {
         return true;
     }
 
+    /**
+     * Returns true if GET request should be routed to primary shards, else false.
+     */
+    protected boolean isPrimaryBasedRouting(ClusterState state, InternalRequest request) {
+        try {
+            if (state.getMetadata()
+                .index(request.concreteIndex())
+                .getSettings()
+                .get(IndexMetadata.SETTING_REPLICATION_TYPE)
+                .equals(ReplicationType.SEGMENT.toString())
+                && request.request().realtime()
+                && request.request().routing() == null
+                && request.request().preference() == null) {
+                return true;
+            }
+        } catch (NullPointerException e) {
+            return false;
```

Comment:
Can we do specific `null` checks here instead of catching the NPE and returning? We might miss edge cases here where the preference will not be applied and we wouldn't necessarily know why. It can become a source of flakiness.